financial-templates-lib/ExpiringMultiPartyEventClient.js

```diff
@@ -0,0 +1,167 @@
+// a clear separation of concerns and to limit the overhead from querying chain necessarily.
+// If no updateThreshold is specified then default to updating every 60 seconds.
+class ExpiringMultiPartyEventClient {
+  constructor(abi, web3, empAddress, updateThreshold = 60) {
+    this.updateThreshold = updateThreshold;
+    this.lastUpdateTimestamp;
+
+    this.web3 = web3;
+
+    // EMP contract
+    this.emp = new web3.eth.Contract(abi, empAddress);
+    this.empAddress = empAddress;
+
+    // EMP Events
+    this.liquidationEvents = [];
+    this.disputeEvents = [];
+    this.disputeSettlementEvents = [];
+  }
+
+  // Calls _update unless it was recently called, as determined by this.updateThreshold.
+  update = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    if (currentTime < this.lastUpdateTimestamp + this.updateThreshold) {
+      Logger.debug({
+        at: "ExpiringMultiPartyEventClient",
+        message: "EMP state update skipped",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTimestamp,
+        timeRemainingUntilUpdate: this.lastUpdateTimestamp + this.updateThreshold - currentTime
+      });
+      return;
+    } else {
+      await this._update();
+      this.lastUpdateTimestamp = currentTime;
+      Logger.debug({
+        at: "ExpiringMultiPartyEventClient",
+        message: "EMP state updated",
+        lastUpdateTimestamp: this.lastUpdateTimestamp
+      });
+    }
+  };
+
+  // Force call of _update, designed to be called by downstream caller that knowingly updated the EMP state.
+  forceUpdate = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    await this._update();
+    this.lastUpdateTimestamp = currentTime;
+    Logger.debug({
+      at: "ExpiringMultiPartyEventClient",
+      message: "EMP state force updated",
+      lastUpdateTimestamp: this.lastUpdateTimestamp
+    });
+  };
+
+  // Delete all events within the client
+  clearState = async () => {
+    this.liquidationEvents = [];
+    this.disputeEvents = [];
+    this.disputeSettlementEvents = [];
+  };
+
+  // Returns an array of liquidation events.
+  getAllLiquidationEvents = () => this.liquidationEvents;
+
+  // Returns an array of dispute events.
+  getAllDisputeEvents = () => this.disputeEvents;
+
+  // Returns an array of dispute events.
+  getAllDisputeSettlementEvents = () => this.disputeSettlementEvents;
+
+  start = () => {
+    this._poll();
+  };
+
+  _poll = async () => {
+    while (true) {
+      try {
+        await this._update();
+      } catch (error) {
+        Logger.error({
+          at: "ExpiringMultiPartyEventClient",
+          message: "client polling error",
+          error: error
+        });
+      }
+      await delay(Number(10_000));
+    }
+  };
+
+  _update = async () => {
+    //Events
+    let liquidationBlockQueryNumber =
+      this.liquidationEvents.length > 0 ? this.liquidationEvents[this.liquidationEvents.length - 1].blockNumber + 1 : 0;
```

Comment:
Maybe instead of setting 0 as the default, we should construct this class with a default starting block. Once we use this on mainnet we wouldn't want to query from the beginning of time right? This would not be a problem if the `getPastEvents()` smart-enough not to look at blocks prior to the contract deployment block, however,

Reply:
ye, GetPastEvents() is fine for this. This is not a bad way to do it and doesn't really make it slower or anything to my understanding. `getPastEvents` optimizes this I think.

Future implementation of this I will feed in the block number as a param in a seperate update function. Something like "update from block x" or "give me events from block x to present". Going to leave this for now as it changes the downstream logic a bit and for now this is the easier way.